server/src/main/java/org/opensearch/common/lucene/search/function/ScriptScoreFunction.java

```diff
@@ -52,8 +52,14 @@
 public class ScriptScoreFunction extends ScoreFunction {
 
     static final class CannedScorer extends Scorable {
-        protected int docid;
-        protected float score;
+        private int docid;
+        private float score;
+
+        public void score(float subScore) {
+            // We check to make sure the script score function never makes a score negative, but we need to make
+            // sure the script score function does not receive negative input.
+            this.score = Math.max(0.0f, subScore);
```

Comment:
@msfroh don't we try to hide the problem instead of fixing the source of negative scoring? As you mentioned, we do catch that in some cases but it slips in other, why cannot we fix that?

Reply:
There was some discussion about this on https://github.com/opensearch-project/OpenSearch/issues/7860.

Essentially, the scoring logic in `cross_fields` match is pretty complicated. Actually, [the Elasticsearch documentation for `cross_fields`](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html#type-cross-fields) even contains the following warning: 

> The cross_fields type blends field statistics in a complex way that can be hard to interpret. The score combination can even be incorrect, in particular when some documents contain some of the search fields, but not all of them.

As @ylwu-amzn called out on the issue, any change we make to the `cross_fields` scoring formula could impact people's "regular" search results. We'd need to be extremely careful.

Meanwhile, this change just turns a query that was throwing an exception (which is what people have complained about) into one that doesn't.

We could possibly wrap the query returned from `MatchQueryBuilder` in a custom scorer that clamps negative values to zero. Functionally that's what happens if `tie_breaker` is 0. I would still include the changes in this PR as an extra layer of protection.

Reply:
@reta  -- I spent some more time today trying to prevent the negative score from bubbling up. So far as I can tell, this is the query that's returning the negative scores: https://github.com/opensearch-project/OpenSearch/blob/4a42150ba0cbf10022797c038f35faaea2f548f6/server/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java#L410

Unfortunately, that just returns a Lucene `DisjunctionMaxQuery`, which normally does the correct thing. The problem arises because of the `adjustDF` and/or the `adjustTTF` methods. Let's pretend that Lucene term stats are something other than what they really are... yay!

I believe it's specifically `adjustTTF` method (called from the `blend` method) that breaks us, because it lowers the `sumTotalTermFreq` to the minimum of `sumTotalTermFreq` across all of the fields, which may be lower than any individual field's term's max `termFreq` (from another field). In theory, the shared `sumTotalTermFreq` should never be lower than the maximum of `termFreq` for any doc across all fields, but we don't have a good way of getting that value without iterating through all documents. I'd love to add a line just before [here](https://github.com/opensearch-project/OpenSearch/blob/4a42150ba0cbf10022797c038f35faaea2f548f6/server/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java#L138) saying `minSumTTF = Math.max(minSumTTF, maxTermFreq)`.

tl;dr: Preventing the negative scores from getting into function/script scoring functions is hard. This `cross_field` scoring is a complicated mess. While it is the root cause, the immediate pain point is *not* the negative scores, but that the negative scores make perfectly valid function/script scoring throw an exception.

I'd like to commit this fix as-is to address the immediate problem. We can address the complicated `cross_field` logic down the road.

Reply:
When we enter the `blend()` method for the integ test I wrote, we start off with `color:red` with `df=3, sumTtf=3`, while `shape:red` has `df=1,sumTtf=1`. Also, `maxDoc` is 3.

1. We exit the first loop with `minSumTTF=2` and `max = 3`. We set `maxDoc` to `2` (since it's bigger than `minSumTTF`). 
2. [Later](https://github.com/opensearch-project/OpenSearch/blob/4a42150ba0cbf10022797c038f35faaea2f548f6/server/src/main/java/org/apache/lucene/queries/BlendedTermQuery.java#L163), we set `actualDf` to 2.

The relevant piece of the explain output is:

```
{
  "value" : -0.22314355,
  "description" : "idf, computed as log(1 + (N - n + 0.5) / (n + 0.5)) from:",
  "details" : [
    {
      "value" : 2,
      "description" : "n, number of documents containing term",
      "details" : [ ]
    },
    {
      "value" : 1,
      "description" : "N, total number of documents with field",
      "details" : [ ]
    }
  ]
},
```

That `n, number of documents containing the term` came from `actualDf`.

I think I can actually fix it by guaranteeing that `actualDf` is never greater than `N`, which I can get from `reader.getDocCount(terms[i].field())`.

So, after saying that fixing the underlying problem is too hard, I think I can fix the underlying problem instead. 😁

Reply:
Here we go: https://github.com/opensearch-project/OpenSearch/pull/13829

It may still be worth merging this PR as an extra layer of protection, but I'd need to update / remove the YAML REST test, because we no longer get a negative score that we force to zero. Alternatively, we could just close this PR, since the underlying problem is (probably) fixed.

Reply:
Thanks a lot for looking into this @msfroh (I learn quite a few things from your clear explanation)